<!DOCTYPE html>
<html lang="ja-JP">
<head>

<meta name="robots" content="noindex">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="../../../style.css">
<link rel="stylesheet" href="../../../syntax-highlight.css">
<link rel="icon" type="image/png" href="../../../rurema.png">

<title>Kernel.#throw (Ruby 3.0.4 リファレンスマニュアル)</title>
<meta name="description" content="[[m:Kernel.#catch]]との組み合わせで大域脱出を行います。 throw
は同じ tag を指定した catch のブロックの終わりまでジャンプします。">
<script src="../../../script.js"></script>
</head>
<body>


<header>
  <nav>
    <ol class="inline-breadcrumb-list">
      <li>
        <a href="../../../doc/index.html">Ruby 3.0.4 リファレンスマニュアル</a>
      </li>
      <li>
        <a href="../../../library/index.html">ライブラリ一覧</a>
      </li>
      <li>
        <a href="../../../library/_builtin.html">組み込みライブラリ</a>
      </li>
      <li>
        <a href="../../../class/-kernel.html">Kernelモジュール</a>
      </li>
      <li>
        throw 
      </li>
    </ol>
  </nav>

<h1>module function Kernel.#throw</h1>
</header>

<main>
<dl class="methodlist">

<dt class="method-heading" id="M_THROW"><code>throw(tag, value = nil) -&gt; ()</code><span class="permalink">[<a href="../../../method/-kernel/m/throw.html">permalink</a>][<a href="https://docs.ruby-lang.org/en/3.0.4/Kernel.html#method-c-throw">rdoc</a>]</span></dt>
<dd class="method-description">
<p>
<a href="../../../method/-kernel/m/catch.html">Kernel.#catch</a>との組み合わせで大域脱出を行います。 throw
は同じ tag を指定した catch のブロックの終わりまでジャンプします。
</p>
<p>
throw は探索時に呼び出しスタックをさかのぼるので、ジャンプ先は同じメソッド内にあるとは限りません。もし ensure節 が存在するならジャンプ前に実行します。
</p>
<p>
同じ tag で待っている catch が存在しない場合は、例外でスレッドが終了します。
</p>
<p>
同じ tag であるとは <a href="../../../method/-object/i/object_id.html">Object#object_id</a> が同じであるという意味です。
</p>
<dl>
<dt class='method-param'>[PARAM] tag:</dt>
<dd>
catch の引数に対応する任意のオブジェクトです。
</dd>
<dt class='method-param'>[PARAM] value:</dt>
<dd>
catch の戻り値になります。
</dd>
<dt>[EXCEPTION] UncaughtThrowError:</dt>
<dd>
同じ tag で待っている catch が存在しない場合に発生します。
</dd>
</dl>
<pre class="highlight ruby">
<span class="caption">例</span>
<code>
<span class="k">def</span> <span class="nf">foo</span>
  <span class="nb">throw</span> <span class="ss">:exit</span>, <span class="mi">25</span>
<span class="k">end</span>

ret <span class="o">=</span> <span class="nb">catch</span><span class="p">(</span><span class="ss">:exit</span><span class="p">)</span> <span class="k">do</span>
  <span class="k">begin</span>
    foo
    some_process<span class="p">(</span><span class="p">)</span>    <span class="c1"># 絶対に実行されない
</span>    <span class="mi">10</span>
  <span class="k">ensure</span>
    <span class="nb">puts</span> <span class="s2">"</span><span class="s2">ensure</span><span class="s2">"</span>
  <span class="k">end</span>
<span class="k">end</span>
<span class="nb">puts</span> ret
<span class="c1">#=&gt; ensure
</span><span class="c1">#   25
</span></code></pre>
<p>
[SEE_ALSO] <a href="../../../method/-kernel/m/catch.html">Kernel.#catch</a>
</p>
</dd>


</dl>
</main>

<footer id="footer">
  <a rel="license" href="https://creativecommons.org/licenses/by/3.0/">
    <img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/3.0/88x31.png" />
  </a>

  <a href="https://github.com/rurema/doctree/issues/new" id="feedback-link">フィードバックを送る</a>
  
  <script>if (window.URLSearchParams) { document.getElementById("feedback-link").search = new URLSearchParams({'body': document.location}); }</script>
</footer>
</body>
</html>
